File: QytetetRuby/lib/calle.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  
  require_relative 'tipo_casilla'
  require_relative'titulo_propiedad'
  class Calle < Casilla
    attr_accessor :numCasas, :numHoteles, :titulo
    def initialize(num_cas,coste,titulo)
      super(num_cas,coste,TipoCasilla::CALLE)
      @numCasas = 0
      @numHoteles = 0
      @titulo = titulo
      
      if @titulo != nil
        @titulo.casilla = self
      end
    end
    
  def asignar_propietario(jugador) 
    @titulo.propietario = jugador
    return @titulo
  end
  
  def calcular_valor_hipoteca
    hipoteca_base = @titulo.hipotecaBase
    cantidad_recibida = hipoteca_base + (@numCasas*0.5) + (@numHoteles*2.0)
    cantidad_recibida;
  end
    
  def cancelar_hipoteca
    @titulo.hipotecado = false
    cantidad_pagar = calcular_valor_hipoteca * 1.1      
    return cantidad_pagar
    
  end

  def cobrar_alquiler    
    coste_alquiler_base = @titulo.alquilerBase
    coste_alquiler = coste_alquiler_base + (@numCasas*0.5)+(@numHoteles*2.0)
    @titulo.cobrar_alquiler(coste_alquiler)
    
    return coste_alquiler    
  end
    
  def edificar_casa
      @numCasas +=1
    
    coste_edificar_casa = @titulo.precioEdificar
    
    return coste_edificar_casa    
  end

  def edificar_hotel
    nuevo_num = @numHoteles + 1
    @numHoteles = nuevo_num
    @numCasas = 0
    coste_edificar_hotel = @titulo.precioEdificar
      
    return coste_edificar_hotel
  end

  def esta_hipotecada
    @titulo.hipotecado
  end

  def get_precio_edificar
    coste_edificar_casa = @titulo.precioEdificar
      
    return coste_edificar_casa
  end

  def hipotecar
    @titulo.hipotecado = true
    cantidad_recibida = calcular_valor_hipoteca
    return cantidad_recibida
  end

  def precio_total_comprar
      precio_compra = @coste + (@numCasas + @numHoteles) * @titulo.precioEdificar
    
    return precio_compra
  end

  def propietario_encarcelado
      encarcelado = @titulo.propietario_encarcelado
    return encarcelado    
  end

  def se_puede_edificar_casa(factor)
    return @numCasas < (4 * factor)
  end

  def se_puede_edificar_hotel(factor)
    return @numCasas == (4*factor) && @numHoteles < (4*factor)
  end

  def soy_edificable
    return true
  end

  def tengo_propietario
    tengo_propietario = @titulo.tengo_propietario
    return tengo_propietario
  end

  def vender_titulo    
    precio_compra = self.precio_total_comprar
    precio_venta = (precio_compra + @titulo.precioEdificar*precio_compra);
        
    @numCasas = 0
    @numHoteles = 0        
    @titulo.propietario= nil        
        
    return precio_venta;    
  end

  def asignar_titulo_propiedad
    
  end
  
  def to_s
        "Número casilla: #{@numeroCasilla}\n Coste: #{@coste}\n Número Hoteles: #{@numHoteles}\n Número Casas: #{@numCasas}\n Tipo: #{@tipo}\n Título propiedad: #{@titulo}"
    end
  
end
end
  

File: modeloqytetet/src/ModeloQytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package ModeloQytetet;

/**
 * 
 * @author pakejo
 */
public class Calle extends Casilla{

    private int coste;
    private int numCasas;
    private int numHoteles;
    private TituloPropiedad titulo;

    public Calle(int numero,int cos,TituloPropiedad titulo) {
        super(numero, TipoCasilla.CALLE);
        this.coste = cos;
        this.numCasas = 0;
        this.numHoteles = 0;
        this.titulo = titulo;
    }
    
    @Override
    public String toString(){
        return "Número: "+super.getNumeroCasilla()+
                "\nNombre: "+titulo.getNombre()+"\nCoste: "+coste+
                "\nNº Casas: "+numCasas+"\nNº Hoteles: "+numHoteles+
                "\nPrecio edificar: "+titulo.getPrecioEdificar()+
                "\nValor hipoteca: "+titulo.getHipotecaBase()+"\n";
    }

    //Consultores
    public int getCoste() {
        return coste;
    }

    public int getNumCasas() {
        return numCasas;
    }

    public int getNumHoteles() {
        return numHoteles;
    }

    public TituloPropiedad getTitulo() {
        return titulo;
    }

    //Modificadores
    public void setNumCasas(int numCasas) {
        this.numCasas = numCasas;
    }

    public void setNumHoteles(int numHoteles) {
        this.numHoteles = numHoteles;
    }
    
    //metodos de instancia
    public void asignarPropietario(Jugador jugador){
        titulo.setPropietario(jugador);
    }
    
    public int calcularValorHipoteca(){
        int hipotecaBase = titulo.getHipotecaBase();
        int cantidadRecibida = (int) (hipotecaBase + (numCasas*0.5) + (numHoteles*2));
        return cantidadRecibida;
    }
    
    public int cancelarHipoteca(){
        titulo.setHipotecada(false);
        int cantidadDar = this.calcularValorHipoteca();
        return cantidadDar;
    }
    
    public int cobrarAlquiler(){
        int costeAlquilerBase = titulo.getAlquilerBase();
        int costeAlquiler = (int) (costeAlquilerBase + (numCasas*0.5) +( numHoteles*2));
        titulo.cobrarAlquiler(costeAlquiler);
        return costeAlquiler;
    }
    
    boolean sePuedeEdificarCasa(){    
        return numCasas < 4*Jugador.getFactorEspeculador();
    }
    
    boolean sePuedeEdificarHotel(){   
        boolean sePuede;
        sePuede = (!this.sePuedeEdificarCasa() && numHoteles <= Jugador.getFactorEspeculador());
        return sePuede;
    }
        
    public int edificarCasa(){
        int costeEdificarCasa = 0;
        
         if(this.sePuedeEdificarCasa()){   
            this.setNumCasas(numCasas+1);
            costeEdificarCasa = titulo.getPrecioEdificar();
         }
        
        return costeEdificarCasa;
    }
    
    public int edificarHotel(){
        
        int costeEdificarHotel = 0;
        
        if(this.sePuedeEdificarHotel()){
            this.setNumHoteles(numHoteles+1);
            costeEdificarHotel = titulo.getPrecioEdificar();
        }
        return costeEdificarHotel;
    }
    
    boolean estaHipotecada(){
        return titulo.getHipotecada();
    }
    
    public boolean tengoPropietario(){
       return titulo.tengoPropietario() == true;
    }
    
    int getPrecioEdificar(){
        return titulo.getPrecioEdificar();
    }
    
    int venderTitulo(){
    
        titulo.setPropietario(null);
        this.setNumHoteles(0);
        this.setNumCasas(0);
        
        int precioCompra = this.getCoste()+(this.getNumCasas()+this.getNumHoteles())*titulo.getPrecioEdificar();
        int precioVenta = (int) (precioCompra+titulo.getFactorRevalorizacion()*precioCompra);
        
        return precioVenta;
    }
    
    public int hipotecar(){
        titulo.setHipotecada(true);
        int cantidadRecibida = this.calcularValorHipoteca();
        return cantidadRecibida;
    }

    @Override
    boolean soyEdificable() {
        return tipo == TipoCasilla.CALLE;
    }    

    boolean propietarioEncarcelado() {
        return titulo.getPropietario().getEncarcelado();
    }
}
